### PR TITLE
[FEATURE] Make the custom dash pattern dialog a style dock widget

### PR DESCRIPTION
**Pull request by nirvn**
## Description
Because it's just really nice to kill one more dialog and see the custom dash pattern reflected on the canvas as the user sets it up:
image

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/gui/symbology/qgsdashspacedialog.h [hunk 1]
```diff
@@ -17,8 +17,31 @@
...
 #include "ui_qgsdashspacedialogbase.h"
+#include "ui_qgsdashspacewidgetbase.h"
+
 #include "qgis_gui.h"
 #include "qgis_sip.h"
+#include "qgspanelwidget.h"
+
+/**
+ * \ingroup gui
+ * A widget to enter a custom dash space pattern for lines
```
Comment (nyalldawson) [inline src/gui/symbology/qgsdashspacedialog.h]: Since 3.8

File: src/gui/symbology/qgsdashspacedialog.h [hunk 2]
```diff
@@ -17,8 +17,31 @@
...
+*/
+class GUI_EXPORT QgsDashSpaceWidget: public QgsPanelWidget, private Ui::QgsDashSpaceWidgetBase
+{
+    Q_OBJECT
+  public:
+
+    //! Constructor for QgsDashSpaceWidget
+    QgsDashSpaceWidget( const QVector<qreal> &v, QWidget *parent SIP_TRANSFERTHIS = nullptr );
```
Comment (nyalldawson) [inline src/gui/symbology/qgsdashspacedialog.h]: Can we change "v" to a full descriptive name?

File: src/gui/symbology/qgsdashspacedialog.h [hunk 3]
```diff
@@ -17,8 +17,31 @@
...
+
+    QVector<qreal> dashDotVector() const;
```
Comment (nyalldawson) [inline src/gui/symbology/qgsdashspacedialog.h]: Dox

File: src/gui/symbology/qgsdashspacedialog.h [hunk 4]
```diff
@@ -34,9 +57,10 @@ class GUI_EXPORT QgsDashSpaceDialog: public QDialog, private Ui::QgsDashSpaceDia
 
     QVector<qreal> dashDotVector() const;
 
-  private slots:
-    void mAddButton_clicked();
-    void mRemoveButton_clicked();
+  private:
+
+    QgsDashSpaceWidget *mWidget;
```
Comment (nyalldawson) [inline src/gui/symbology/qgsdashspacedialog.h]: = nullptr

**Comment by nirvn** [comment]
@nyalldawson , comments addressed, including google hangout suggestions to i/ add a unit type hint in the column headers, and ii/ rework the placement of the push button to harmonize the UI with other parts.

File: src/core/qgsunittypes.cpp [hunk 5]
```diff
@@ -186,6 +186,39 @@ QString QgsUnitTypes::toString( DistanceUnit unit )
...
+      return QObject::tr( "pt", "render" );
+
+    case RenderInches:
+      return QObject::tr( "in", "render" );
+
+    case RenderUnknownUnit:
+      return QObject::tr( "unknown", "render" );
+
+    case RenderMetersInMapUnits:
+      return QObject::tr( ",", "render" );
```
Comment (nyalldawson) [inline src/core/qgsunittypes.cpp]: Comma?

**Review by nyalldawson — COMMENTED** [review]
Better add some tests for the new unit method, but otherwise looks great!

File: src/core/qgsunittypes.cpp (hunk 5 above)
Comment (nirvn) [reply src/core/qgsunittypes.cpp]: Doh! For the record, the character "," is right next to "m" on my keyboard :wink:

**Pull request merged**